gunicorn.conf.py

```diff
@@ -21,11 +25,34 @@ def post_fork(server, worker):
     tracing.setup_default_tracing()
 
 
-# this hook is called when the gunicorn worker receives a SIGABRT from the
-# master, because it has timed out. By default, it just kills the worker.
-# However, we raise an Exception, so that our django stack will handle it as an
-# error and emit logs and telemetry.
+# track this worker is currently handling an actual request
+def pre_request(worker, req):
+    global _http_request
+    _http_request = True
+
+
+# track this worker is no longer handling a request
+def post_request(worker, req, environ, resp):
+    global _http_request
+    _http_request = False
+
+
+# This hook is called when the gunicorn worker receives a SIGABRT from the
+# master, because it has timed out. By default, it just kills the worker, we do
+# something different.
+#
+# Firstly, we check if we are currently handing an acutal HTTP request. If not,
+# it means we just got a connection with no data, so we exit w/o error. This
+# happens with airlock specifically as we have no proxy infront of it.
+#
+# If not, we raise RequestTimeout, so that our django stack will handle it as
+# an error and emit logs and telemetry.
 def worker_abort(worker):
+    if not _http_request:
+        worker.log.info(f"No request sent timeout, exiting (pid: {worker.pid})")
+        # do not raise SystemExit, just quit
+        os._exit(0)
```

Comment:
Is there a change in behaviour in exiting 0 vs 1?

Reply:
Yes and no :)

See the last paragraph of this commit from the initial introduction of raising RequestTimeout: https://github.com/opensafely-core/airlock/commit/b61eece0948bbda7c731d97c47eeaaa2e850c285

Basically, because we raise [at this point](https://github.com/benoitc/gunicorn/blob/master/gunicorn/workers/base.py#L203), then the worker actually drops out of this loop [at this point](https://github.com/benoitc/gunicorn/blob/master/gunicorn/workers/sync.py#L61) (as self.worker is now False), and exits with a 0.

So, yes, it's a change from the default gunicorn behaviour, but no this commit is not a change, we already made that change in the earlier commit.

Reply:
It would actually be quite complex to *keep* the behaviour of exiting with a 1 when raising an exception.

Of course, we could choose to exit with ever exit code we want here. Because there's no error in this case, there's nothing to see, so I figure 0 was best?